model-datastructure/src/commonMain/kotlin/org/modelix/model/VersionMerger.kt

```diff
@@ -45,11 +45,32 @@ class VersionMerger(private val storeCache: IDeserializingKeyValueStore, private
         }
     }
 
+    private fun collectLatestNonMerges(version: CLVersion?, visited: MutableSet<String>, result: MutableSet<Long>) {
+        if (version == null) return
+        if (!visited.add(version.getContentHash())) return
+        if (version.isMerge()) {
+            collectLatestNonMerges(version.getMergedVersion1(), visited, result)
+            collectLatestNonMerges(version.getMergedVersion2(), visited, result)
+        } else {
+            result.add(version.id)
+            collectLatestNonMerges(version.baseVersion, visited, result)
+        }
+    }
+
     protected fun mergeHistory(leftVersion: CLVersion, rightVersion: CLVersion): CLVersion {
         if (leftVersion.hash == rightVersion.hash) return leftVersion
         val commonBase = Companion.commonBaseVersion(leftVersion, rightVersion)
         if (commonBase?.hash == leftVersion.hash) return rightVersion
         if (commonBase?.hash == rightVersion.hash) return leftVersion
+
+        val leftNonMerges = HashSet<Long>().also { collectLatestNonMerges(leftVersion, HashSet(), it) }
+        val rightNonMerges = HashSet<Long>().also { collectLatestNonMerges(rightVersion, HashSet(), it) }
+        if (leftNonMerges == rightNonMerges) {
+            // If there is no actual change on both sides, but they just did the same merge, we have to pick one
+            // of them, otherwise both sides will continue creating merges forever.
+            return if (leftVersion.id < rightVersion.id) leftVersion else rightVersion
```

Comment:
Maybe a detailed and easier to understand unit test for exactly this condition and function would be a good addition to the test code base. The ReplicatedRepositoryTest is pretty complex and maybe there's a more targeted version that also helps documenting the corner-case for others reading the code.

Reply:
Maybe in a separate PR.